sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,358 @@
+       * and the other is a literal.
+       * Note that: all binary predicate computing methods assume the literal is at the right side,
+       * so we will change the predicate order if not.
+       */
+      case op@LessThan(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@LessThan(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(GreaterThan(ar, l), planStat, ar, l, update)
+
+      case op@LessThanOrEqual(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@LessThanOrEqual(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(GreaterThanOrEqual(ar, l), planStat, ar, l, update)
+
+      case op@GreaterThan(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@GreaterThan(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(LessThan(ar, l), planStat, ar, l, update)
+
+      case op@GreaterThanOrEqual(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@GreaterThanOrEqual(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(LessThanOrEqual(ar, l), planStat, ar, l, update)
+
+      /** EqualTo does not care about the order */
+      case op@EqualTo(ExtractAttr(ar), l: Literal) =>
+        evaluateBinary(op, planStat, ar, l, update)
+      case op@EqualTo(l: Literal, ExtractAttr(ar)) =>
+        evaluateBinary(op, planStat, ar, l, update)
+
+      case In(ExtractAttr(ar), expList) if !expList.exists(!_.isInstanceOf[Literal]) =>
+        /**
+         * Expression [In (value, seq[Literal])] will be replaced with optimized version
+         * [InSet (value, HashSet[Literal])] in Optimizer, but only for list.size > 10.
+         * Here we convert In into InSet anyway, because they share the same processing logic.
+         */
+        val hSet = expList.map(e => e.eval())
+        evaluateInSet(planStat, ar, HashSet() ++ hSet, update)
+
+      case InSet(ExtractAttr(ar), set) =>
+        evaluateInSet(planStat, ar, set, update)
+
+      case Like(_, _) | Contains(_, _) | StartsWith(_, _) | EndsWith(_, _) =>
+        evaluateLike(condition, planStat, update)
+
+      // TODO: it's difficult to estimate IsNull after outer joins
+      // case IsNull(ExtractAttr(ar)) =>
+      // evaluateIsNull(planStat, ar, update)
+      case IsNotNull(ExtractAttr(ar)) =>
+        evaluateIsNotNull(planStat, ar, update)
+
+      case _ =>
+        logDebug("[CBO] Unsupported filter condition: " + condition)
+        notSupported = true
+        1.0
+    }
+    if (notSupported) {
+      1.0
+    } else if (isNot) {
+      1.0 - percent
+    } else {
+      percent
+    }
+  }
+
+  def evaluateIsNotNull(
+      planStat: Statistics,
+      attrRef: AttributeReference,
+      update: Boolean)
+    : Double = {
+    if (!planStat.colStats.contains(attrRef.name)) {
+      logInfo("[CBO] No statistics for " + attrRef)

```

Comment:
logInfo is fine.  This is because this method may still get executed even if a user does not collect statistics on a given column.